### ISSUE TITLE
feasibility of providing as a simple tensorflow model

### ISSUE THREAD

**Issue by jerabaul29** [google/lyra#75]
I am curious about whether this could be provided as a simple tensorflow model. I.e. something where one could just:

```python
lyra_encoder = keras.models.load_model('path_to_lyra_encoder_model')
lyra_decoder = keras.models.load_model('path_to_lyra_decoder_model')
```

and from there one would be able to simply

```python
encoded_frame = lyra_encoder.predict(frame)
decode_frame = lyra_decoder.predict(encoded_frame)
```

Releasing lyra in such a way would make it so much easier to use on a wide range of environments, from microcontrollers to gpus.

Is there any fundamental reason for such a way to release to not work, or "is it just not how it is done as of today"?

**Comment by aluebs** [comment]
Today only the generative model at the decoder is trained in TensorFlow, so it would be quite a bit of work to port the rest over.
I can see the value of having TF models like that, but at least for some use-cases it is preferred to have a C++ API to integrate with.

**Comment by jerabaul29** [comment]
Temsoflow has a C++ API too though :) . Actually, this would be one of the use I would like to try - using this on a moder Cortex MCU, through the tensorflow lite C++ API.

**Comment by aluebs** [comment]
That is a great point. But that would require a large effort, since big parts of our system aren't implemented in TensorFlow today.

**Comment by jerabaul29** [comment]
A few thoughts; just brainstorming, may be stupid ideas:

- since the architecture of the decoder and encoder sides are well known, I guess they should be easy to implement in Keras? Or is there anything unusual about these?

- are the training data open with pipelines available to generate these? If yes, training anew with the same network in Keras and same data would be easy. If not, maybe some other voice dataset could be used, at least first?

If not these points are easy to implement, wonder if it would be possible to train a Keras model to match the present encoder decoder pair, simply by generating large amounts of synthetic data using Lyra and using these as training data for the Keras version.

**Comment by aluebs** [comment]
Those are definitely all great points.
Regarding re-implementing in Keras, it would definitely be some work. But on top of that, we are currently working on [Lyra v2](https://ai.googleblog.com/2021/08/soundstream-end-to-end-neural-audio.html), which will have many improvements and changes. We are planning on releasing it here as soon as it is ready, so I would wait before putting in the effort.
And regarding the training data, we are not planning on releasing that. So if someone would want to re-train the system, it would need to be using some other available dataset.
I don't think there is much value to train on samples generated synthetically by Lyra, since it would only introduce additional artifacts.

**Comment by jerabaul29** [comment]
Ok, thanks!

- I agree training on synthetic data would just introduce artifacts, but as a first quick fix it would be better than no tensorflow model :) 

- Sounds exciting with Lyra V2; if it can be available as a tensorflow model that would be awesome :) .

**Comment by aluebs** [comment]
I still think it would be better to train on clean (and noisy) recordings without having gone through Lyra. The current feature extraction is traditional, not learned, so it could be written out with no training.

And just for transparency sake, there is no plan to provide an encoder and decoder TensorFlow models at the moment.

**Comment by rocco-haro** [comment]
Thanks for making this open source, really appreciate your contributions @aluebs. 

Definitely a +1 on the enc/dec w/ TF :) 

Excited for the release of v2, approximately when can we expect it?

**Comment by aluebs** [comment]
There is no set date for the Lyra v2 release, but it should happen very soon (within the next month or two).
It will still have the same C++ API, but, because it is trained end-to-end, many components are migrated to TFLite.

**Comment by aluebs** [comment]
Yes, one advantage of migrating to TFLite is the portability.
There will still be fair amount of C++ logic and glue, so full tensorflow implementation might require some porting work.

**Comment by wangtiance** [comment]
It's great to know that Lyra v2 is coming out! Will you provide a training pipeline (so that we can train and tweak the model using our own data), or just a pretrained model?

**Comment by aluebs** [comment]
At this time we'll only open-source a pretrained model for inference.

**Comment by aluebs** [comment]
Since it was mentioned in this discussion, [Lyra 1.2.0](https://github.com/google/lyra/releases/tag/v1.2.0) has been release, together with its TFLite models.

**Comment by jerabaul29** [comment]
So this is effectively done, right? :) Feel free to close.

**Comment by aluebs** [comment]
The TFLite models are still divided in encoder, quantizer and decoder and I am not sure how simple it would be to implement something along the lines of the initial suggestion. So I am not sure that it is done, but at least it is a step in the right direction. I'd love to know if someone implements python wrappers like that.
